cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,307 @@
+import co.cask.cdap.internal.io.DatumReader;
+import co.cask.cdap.internal.io.DatumReaderFactory;
+import co.cask.cdap.internal.io.SchemaGenerator;
+import co.cask.cdap.messaging.MessageFetcher;
+import co.cask.cdap.messaging.MessagingService;
+import co.cask.cdap.messaging.MessagingUtils;
+import co.cask.cdap.messaging.data.RawMessage;
+import co.cask.cdap.metrics.store.MetricDatasetFactory;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.reflect.TypeToken;
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.name.Named;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Process metrics by consuming metrics being published to TMS.
+ */
+public class MessagingMetricsProcessorService extends AbstractMetricsProcessorService {
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingMetricsProcessorService.class);
+
+  private static final byte[] INITIAL_LAST_MESSAGE_ID = Bytes.toBytes(-1L);
+
+  private final TopicId[] metricsTopics;
+  private final TopicIdMetaKey[] metricsTopicMetaKeys;
+  private final MessagingService messagingService;
+  private final DatumReader<MetricValues> recordReader;
+  private final Schema recordSchema;
+  private final MetricStore metricStore;
+  private Map<String, String> metricsContextMap;
+  private List<ProcessMetricsThread> processMetricsThreadList;
+  private final int fetcherPersistThreshold;
+
+  private long lastLoggedMillis;
+  private long recordsProcessed;
+
+  @Inject
+  public MessagingMetricsProcessorService(MetricDatasetFactory metricDatasetFactory,
+                                          @Named(Constants.Metrics.TOPIC_PREFIX) String topicPrefix,
+                                          @Assisted Set<Integer> partitions,
+                                          MessagingService messagingService,
+                                          SchemaGenerator schemaGenerator,
+                                          DatumReaderFactory readerFactory,
+                                          MetricStore metricStore,
+                                          @Named(Constants.Metrics.MESSAGING_FETCHER_PERSIST_THRESHOLD)
+                                            int fetcherPersistThreshold) {
+    super(metricDatasetFactory);
+    this.metricsTopics = new TopicId[partitions.size()];
+    this.metricsTopicMetaKeys = new TopicIdMetaKey[partitions.size()];
+    for (int i = 0; i < partitions.size(); i++) {
```

Comment:
Do you mean there can be multiple processors reading from different ranges of topics? However, it seems there's always only one metrics processor started by `MetricsProcessorTwillRunnable`